README.md
# LRTimelapse-Pro-Timer-Free

Project Infos on:
English:
https://lrtimelapse.com/lrtimelapse-pro-timer-free/

German:
http://gwegner.de/know-how/lrtimelapse-pro-timer-free/




releasenotes.txt
0.92
Improvements by Hans Vollmer
- Timing for Cam release and exposure indicator in timer Interrupt, see https://forum.lrtimelapse.com/Thread-lrtimelapse-pro-timer-improvements

0.91
- Added 1 sec decoupling to timelapse and single bulb shots. This allows the photographer to take the hands from the timer before the shooting starts.
You can configure the time in the sketch (const decouple time in line 37)

0.90
 -Intervals can now be a maximum of 999 secs (Over 16 Minutes) instead of 99 secs only.
 -Added acceleration when changing the intervals: intervals higher then 20secs will get increased/decreased by one second, intervals below this by 0.1.
 
0.89
Changes by Klaus Heiss, Austria
- Introduces LCD backlight dimming for power saving and optical reason 
  long pressing select key steps dimming level. release key at suitable level
  when running: up down keys are increasing or decreasing brighness for your convenience
  consumption at available brightness leveles to get an idea of possible power savings:
  0 (OFF) 	 50 mA
  1 min		 55 mA  (you may now let backlight on always at this level)
  2		 70 mA
  3		 88 mA
  4		104 mA
  5		115 mA
  6		124 mA

- Params can be saved in EEPROM
  when backlight is turned on and pressing select key 2 times short and fast (double-pressing) these params are saved in EEPROM
  - actual display background brightness level
  - actual interval 
  After power on or reset these values are restored from EEPROM
  
- Turned OFF Onboard LED 13. it only consumes battery power.

0.88
Changes by Klaus Heiss, Austria (Thank you!)
- Introduces dynamic key repeating. This makes the keyboard more responsive.
  Key repetition starts with 2 per second and then increases to 10 keystrokes per second.
- changed Menu text for "No of shots" form "----" to "unlimited".  
- Made the key recognition less sensitive, so it will work without changes with a wider variety of displays.
- Added alternative key values into LCD_Keypad_Reader.cpp for a similar looking Display from Sainmart, but with other voltage steps. You may comment out this alternative values, if the key pressures are not properly recognized on your board.

0.87
- Added Left click Single Exposure screen. From the main (interval) screen press left, to reach the Single Exposure screen. Here you can either briefly trigger a single shot from the camera (press left) or you can use the up/down keys to set a bulb time.
Thanks to user Colin that brought up the idea in his fork on https://github.com/ColMol/LRTimelapse-Pro-Timer-Free/tree/master/LRTimelapse_Pro_Timer_Free)

0.86
- Added Shooting-Mode menu, M vs. Bulb
- New Bulb timelapse shooting mode for astro-photographers that would like to define the exposure time via the timer: in M-Mode, the trigger time by the timer will be the shortest possible for regular timelapse shooting, exposure will be set in the camera. In Bulb-Mode the exposure will be controlled and can be defined in the timer. The maximum possible exposure will be Interval - 0.5 secs dark time. This dark-time is defined as constant in the source code and can be changed, if necessary.
- Improved key responsiveness

0.85 first public version

LRTimelapse_Pro_Timer_Free/EEPROMConfig.h
/*
   Klaus Heiss
   www.elite.at
   Aug. 2017
*/

#ifndef __EEPROM_Config__
#define __EEPROM_Config__

#include <Arduino.h>

// #define debug
#undef debug

//                                        N/A   BGL:0..5  interval   N/A
#define EEPParamsDefault (EEPParamsStru) { 0,          4,      4.0,   0 };

struct EEPParamsStru {
  unsigned int SumBytes;                      // 2
  byte         BackgroundBrightnessLevel;     // 1
  float        Interval;                      // 4
  unsigned int SumBytesXored;                 // 2
};                                            // 9 Bytes

class EEPParams {
  private:
    unsigned int CRCSumBytes;
    boolean crcOK();
    void SerialPrintParams();
  public:
    EEPParamsStru Params;
    boolean ParamsRead();
    boolean ParamsWrite();
    
};

#endif


LRTimelapse_Pro_Timer_Free/EEPROMConfig.cpp
/*
   Klaus Heiss
   www.elite.at
   Aug. 2017
*/

#include "EEPROMConfig.h"
#include <EEPROM.h>

void EEPParams::SerialPrintParams() {
  #ifdef debug
    String str = "";
    str = "Background brightness level = "; str += Params.BackgroundBrightnessLevel;  Serial.println(str);
    str = "Interval                    = "; str += Params.Interval;  Serial.println(str);
  #endif
}

boolean EEPParams::ParamsWrite() {
  #ifdef debug
    Serial.println("=== Write Params to EEPROM if changed ===");
  #endif
  EEPParamsStru epp;
  EEPROM.get(0,epp);
  if ( (Params.SumBytes                  != epp.SumBytes) ||    
       (Params.BackgroundBrightnessLevel != epp.BackgroundBrightnessLevel) ||
       (Params.Interval                  != epp.Interval) ||
       (Params.SumBytesXored             != epp.SumBytesXored) ) {
    crcOK();  // calcs & sets CRCSumBytes
    Params.SumBytes =      CRCSumBytes;
    Params.SumBytesXored = CRCSumBytes^0xFFFF;
    EEPROM.put(0,Params);
    #ifdef debug
      SerialPrintParams();
      Serial.println("wrote Params into EEPROM.");
    #endif
    return true;
  } else {
    #ifdef debug
      Serial.println("params didn't change.");
    #endif
    return false;
  }
}

boolean EEPParams::ParamsRead() {
  #ifdef debug
    Serial.println("=== Read Params from EEPROM ===");
  #endif
  // read Params fromEEPROM
  EEPROM.get(0,Params); 
  boolean retval = crcOK(); 
  if (retval) {
    #ifdef debug
      Serial.println("EEPROM Params (crc) OK");
    #endif
  } else {
    SerialPrintParams();
    Params = EEPParamsDefault;
    #ifdef debug
      Serial.println("EEPROM Params CRC WRONG - uses defaults");
    #endif
  }
  SerialPrintParams();
  return retval;
}

boolean EEPParams::crcOK() {
  int len = sizeof(Params);     //12abbbb12    len=9 -4 = 5
  byte * pb;
  pb = (byte *) & Params;
  pb += 2;
  CRCSumBytes = 0;
  for (int i=0; i < (len-4); i++) {
    CRCSumBytes += *pb++;
  }
  #ifdef debug
    Serial.println("=== Calc Params crc ===");
    String str = "";
    str = "Size Of Params= "; str += len;  Serial.println(str);
    
    str = "Params.SumBytes      = "; str += String(Params.SumBytes, HEX); Serial.println(str);
    str = "Calced SumBytes      = "; str += String(CRCSumBytes, HEX); Serial.println(str);
    
    str = "Params.SumXORedBytes = "; str += String(Params.SumBytesXored, HEX); Serial.println(str);
    str = "Calced SumXORedBytes = "; str += String(CRCSumBytes^0xFFFF, HEX); Serial.println(str);
  #endif
  boolean res = (CRCSumBytes == Params.SumBytes) && ((CRCSumBytes^0xFFFF) == Params.SumBytesXored);
  return res;
}



LRTimelapse_Pro_Timer_Free/LCD_Keypad_Reader.h
/*
  Pro-Timer Free
  Gunther Wegner
  http://gwegner.de
  http://lrtimelapse.com
  https://github.com/gwegner/LRTimelapse-Pro-Timer-Free

  Version 0.88: Thanks for Klaus Heiss (KH) for implementing the dynamic key rate
*/

#ifndef LCD_Keypad_Reader_h
#define LCD_Keypad_Reader_h

#include "Arduino.h"

#define SAMPLE_WAIT -1
#define NO_KEY 0
#define UP_KEY 3
#define DOWN_KEY 4
#define LEFT_KEY 2
#define RIGHT_KEY 5
#define SELECT_KEY 1

class LCD_Keypad_Reader
{
  public:
    LCD_Keypad_Reader();
    int getKey();
    int categorizeKey(int);
    int RepeatRate;
    int ActRepeatRate();
  private:
    int _keyPin;
    int _threshold;
    int _curInput;
    int _curKey;
};

const  int keySampleRate          = 100;      // ms between checking keypad for key

/* K.H: "Added KeyRepeatRateAcceleration" to   customize dynamic behavior */
const  int keyRepeatRateHighDelay = 500;      // delay for speeding up
const  int keyRepeatRateStep      =  60;      // speeding up steps
const  int keyRepeatRateSlow      = 500;      // SLOW speed when held, key repeats 1000 / keyRepeatRate times per second
const  int keyRepeatRateHigh      = 100;      // HIGH speed when held, key repeats 1000 / keyRepeatRate times per second


#endif


LRTimelapse_Pro_Timer_Free/.vs/__vm/.vsarduino.h
/* 
	Editor: http://www.visualmicro.com
			visual micro and the arduino ide ignore this code during compilation. this code is automatically maintained by visualmicro, manual changes to this file will be overwritten
			the contents of the Visual Micro sketch sub folder can be deleted prior to publishing a project
			all non-arduino files created by visual micro and all visual studio project or solution files can be freely deleted and are not required to compile a sketch (do not delete your own code!).
			note: debugger breakpoints are stored in '.sln' or '.asln' files, knowledge of last uploaded breakpoints is stored in the upload.vmps.xml file. Both files are required to continue a previous debug session without needing to compile and upload again
	
	Hardware: Arduino/Genuino Uno, Platform=avr, Package=arduino
*/

#if defined(_VMICRO_INTELLISENSE)

#ifndef _VSARDUINO_H_
#define _VSARDUINO_H_
#define __AVR_ATmega328p__
#define __AVR_ATmega328P__
#define F_CPU 16000000L
#define ARDUINO 10609
#define ARDUINO_AVR_UNO
#define ARDUINO_ARCH_AVR
#define __cplusplus 201103L
#define __AVR__
#define __inline__
#define __asm__(...)
#define __extension__
#define __inline__
#define __volatile__
#define GCC_VERSION 40902

#define __cplusplus 201103L
#undef __cplusplus
#define __cplusplus 201103L

#define volatile(va_arg) 
#define _CONST
#define __builtin_va_start
#define __builtin_va_end
#define __attribute__(...)
#define NOINLINE __attribute__((noinline))
#define prog_void
#define PGM_VOID_P int


#ifndef __builtin_constant_p
	#define __builtin_constant_p __attribute__((__const__))
#endif
#ifndef __builtin_strlen
	#define __builtin_strlen  __attribute__((__const__))
#endif


#define NEW_H
typedef void *__builtin_va_list;
//extern "C" void __cxa_pure_virtual() {;}

typedef int div_t;
typedef int ldiv_t;


typedef void *__builtin_va_list;
//extern "C" void __cxa_pure_virtual() {;}



#include <Arduino.h>
#include <pins_arduino.h> 
#undef F
#define F(string_literal) ((const PROGMEM char *)(string_literal))
#undef PSTR
#define PSTR(string_literal) ((const PROGMEM char *)(string_literal))


#define pgm_read_byte(address_short) uint8_t() 
#define pgm_read_word(address_short) uint16_t() 
#define pgm_read_dword(address_short) uint32_t()
#define pgm_read_float(address_short) float()
#define pgm_read_ptr(address_short)   short()

#include "LRTimelapse_Pro_Timer_Free.ino"
#endif
#endif


LRTimelapse_Pro_Timer_Free/LCD_Keypad_Reader.cpp
/*
  Pro-Timer Free
  Gunther Wegner
  http://gwegner.de
  http://lrtimelapse.com
  https://github.com/gwegner/LRTimelapse-Pro-Timer-Free

  Version 0.88: Thanks for Klaus Heiss (KH) for implementing the dynamic key rate
*/

#include "Arduino.h"
#include "LCD_Keypad_Reader.h"

static int DEFAULT_KEY_PIN = 0;

// The Sainsmart keypad uses a voltage divider to deliver a voltage
// between 0 and 5 V that corresponds to the key being pressed in
// order to use only a single input pin. The values below are from 0 to
// 1023 because the Arduino uses a 10 bit resolution.
/* Board introduced by G.W.   */
static int DEFAULT_THRESHOLD = 50;   // KH: changed from 5 to 50 - more tolerance
static int UPKEY_ARV = 144; // 0.720 V, that's read "analogue read value"
static int DOWNKEY_ARV = 329; // 1.645 V
static int LEFTKEY_ARV = 505; // 2.525 V
static int RIGHTKEY_ARV = 0; // 0 V
static int SELKEY_ARV = 742; // 3.710 V
static int NOKEY_ARV = 1023; // 5.115 V
/* */

/* K.H.: My NEW Board (looks same as that one shown in project, but branded as: Open Source ELECTRONICS - Miroad
 *       has other voltage steps - here the values:    */
/*  IF ONLY THE RIGHT-KEY IS RECOGNIZED ON YOUR BOARD,  TRY THIS VALUES
static int DEFAULT_THRESHOLD = 50;
static int UPKEY_ARV = 99;
static int DOWNKEY_ARV = 256;
static int LEFTKEY_ARV = 410;
static int RIGHTKEY_ARV = 0;
static int SELKEY_ARV = 640;
static int NOKEY_ARV = 1023;
*/

LCD_Keypad_Reader::LCD_Keypad_Reader()
{
  _keyPin = DEFAULT_KEY_PIN;
  _threshold = DEFAULT_THRESHOLD;
  _curInput = NO_KEY;
  _curKey = NO_KEY;
  RepeatRate = keyRepeatRateSlow;
}

int LCD_Keypad_Reader::ActRepeatRate()
{
  RepeatRate -= keyRepeatRateStep;
  if (RepeatRate < keyRepeatRateHigh){
    RepeatRate =keyRepeatRateHigh;
  }
  return RepeatRate;
}

int LCD_Keypad_Reader::getKey()
{
  _curInput =  analogRead(_keyPin);
  _curKey = categorizeKey(_curInput);
  return _curKey;
}

int LCD_Keypad_Reader::categorizeKey(int analogKeyValue){
  int categorizedKeyValue = 0;

  if (analogKeyValue > UPKEY_ARV - _threshold && analogKeyValue < UPKEY_ARV + _threshold ){
      categorizedKeyValue = UP_KEY;
  }
  else if (analogKeyValue > DOWNKEY_ARV - _threshold && analogKeyValue < DOWNKEY_ARV + _threshold ){
      categorizedKeyValue = DOWN_KEY;
  }
  else if (analogKeyValue > RIGHTKEY_ARV - _threshold && analogKeyValue < RIGHTKEY_ARV + _threshold ){
      categorizedKeyValue = RIGHT_KEY;
  }
  else if (analogKeyValue > LEFTKEY_ARV - _threshold && analogKeyValue < LEFTKEY_ARV + _threshold ){
      categorizedKeyValue = LEFT_KEY;
  }
  else if (analogKeyValue > SELKEY_ARV - _threshold && analogKeyValue < SELKEY_ARV + _threshold ){
      categorizedKeyValue = SELECT_KEY;
  }
  else{
    categorizedKeyValue = NO_KEY;
  }

  return categorizedKeyValue;
}
